src/main/java/edu/wpi/grip/core/Step.java

```diff
@@ -78,13 +78,25 @@ protected Pipeline getPipeline() {
         return this.pipeline;
     }
 
+    private synchronized void runPerformIfPossible() {
+        for (InputSocket<?> inputSocket : inputSockets) {
```

Comment:
Definitely comment this to explain why we do this (I know the answer, but I don't think it's immediately obvious)